File: app/src/main/java/ch/unstable/ost/error/ErrorUtils.java
package ch.unstable.ost.error;


import android.content.pm.Signature;
import android.util.Log;
import android.view.View;

import androidx.annotation.StringRes;

import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.cert.CertificateEncodingException;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;

import ch.unstable.ost.BuildConfig;
import ch.unstable.ost.utils.NavHelper;
import ch.unstable.ost.R;

public enum ErrorUtils {
    ;
    private static final String TAG = "ErrorUtils";

    /*
    * Get the X.509 certificate.
    */
    private static X509Certificate getCertificateFromSignature(Signature signature) throws CertificateException {
        final byte[] rawCert = signature.toByteArray();
        InputStream certStream = new ByteArrayInputStream(rawCert);
        CertificateFactory certFactory = CertificateFactory.getInstance("X509");
        return (X509Certificate) certFactory.generateCertificate(certStream);
    }

    public static byte[] getSHA256Fingerprint(X509Certificate certificate) throws CertificateEncodingException {
        try {
            return getFingerprint(certificate, "SHA-256");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    private static byte[] getFingerprint(X509Certificate cert, String algorithm)
            throws NoSuchAlgorithmException, CertificateEncodingException {
        MessageDigest md = MessageDigest.getInstance(algorithm);
        byte[] der = cert.getEncoded();
        md.update(der);
        return md.digest();

    }

    public static String printableSignature(Signature signature) {
        X509Certificate certificate;
        try {
            certificate = getCertificateFromSignature(signature);
        } catch (Exception e) {
            return signature.toCharsString();
        }
        StringBuilder builder = new StringBuilder();
        builder.append("Subject: ");
        builder.append(certificate.getSubjectDN().getName());
        builder.append('\n');
        builder.append("Issuer: ");
        builder.append(certificate.getIssuerDN().getName());
        builder.append('\n');
        builder.append("Serial Number: ");
        builder.append(certificate.getSerialNumber());
        builder.append('\n');
        builder.append("Not After: ");
        builder.append(certificate.getNotAfter());
        builder.append('\n');
        builder.append("Not Before: ");
        builder.append(certificate.getNotBefore());
        builder.append('\n');
        return builder.toString();
    }

    public static void showErrorSnackbar(View view, @StringRes int errorMessage, final Throwable throwable) {
        if (BuildConfig.DEBUG) {
            Log.e(TAG, view.getContext().getString(errorMessage), throwable);
        }
        Snackbar.make(view, errorMessage, Snackbar.LENGTH_LONG)
                .setAction(R.string.action_resport_error, view1 -> NavHelper.INSTANCE.startErrorActivity(view1.getContext(), throwable))
                .show();
    }
}

File: app/src/main/java/ch/unstable/ost/error/ErrorUtils.kt
package ch.unstable.ost.error

import android.content.pm.Signature
import android.util.Log
import android.view.View
import androidx.annotation.StringRes
import ch.unstable.ost.BuildConfig
import ch.unstable.ost.R
import ch.unstable.ost.utils.NavHelper.startErrorActivity
import com.google.android.material.snackbar.Snackbar
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.security.cert.CertificateEncodingException
import java.security.cert.CertificateException
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate

object ErrorUtils {
    private const val TAG = "ErrorUtils"

    /*
    * Get the X.509 certificate.
    */
    @Throws(CertificateException::class)
    private fun getCertificateFromSignature(signature: Signature): X509Certificate {
        val rawCert = signature.toByteArray()
        val certStream: InputStream = ByteArrayInputStream(rawCert)
        val certFactory = CertificateFactory.getInstance("X509")
        return certFactory.generateCertificate(certStream) as X509Certificate
    }

    @Throws(CertificateEncodingException::class)
    fun getSHA256Fingerprint(certificate: X509Certificate): ByteArray {
        return try {
            getFingerprint(certificate, "SHA-256")
        } catch (e: NoSuchAlgorithmException) {
            throw RuntimeException(e)
        }
    }

    @Throws(NoSuchAlgorithmException::class, CertificateEncodingException::class)
    private fun getFingerprint(cert: X509Certificate, algorithm: String): ByteArray {
        val md = MessageDigest.getInstance(algorithm)
        val der = cert.encoded
        md.update(der)
        return md.digest()
    }

    fun printableSignature(signature: Signature): String {
        val certificate: X509Certificate = try {
            getCertificateFromSignature(signature)
        } catch (e: Exception) {
            return signature.toCharsString()
        }
        val builder = StringBuilder()
        builder.append("Subject: ")
        builder.append(certificate.subjectDN.name)
        builder.append('\n')
        builder.append("Issuer: ")
        builder.append(certificate.issuerDN.name)
        builder.append('\n')
        builder.append("Serial Number: ")
        builder.append(certificate.serialNumber)
        builder.append('\n')
        builder.append("Not After: ")
        builder.append(certificate.notAfter)
        builder.append('\n')
        builder.append("Not Before: ")
        builder.append(certificate.notBefore)
        builder.append('\n')
        return builder.toString()
    }

    @JvmStatic
    fun showErrorSnackbar(view: View, @StringRes errorMessage: Int, throwable: Throwable?) {
        if (BuildConfig.DEBUG) {
            Log.e(TAG, view.context.getString(errorMessage), throwable)
        }
        Snackbar.make(view, errorMessage, Snackbar.LENGTH_LONG)
                .setAction(R.string.action_resport_error) { view1: View -> startErrorActivity(view1.context, throwable!!) }
                .show()
    }
}